Readme update 06/2020 (#799)
microsoft/vscode-azurearmtools: README.md

[![Version](https://vsmarketplacebadge.apphb.com/version/msazurermtools.azurerm-vscode-tools.svg)](https://marketplace.visualstudio.com/items?itemName=msazurermtools.azurerm-vscode-tools) [![Installs](https://vsmarketplacebadge.apphb.com/installs-short/msazurermtools.azurerm-vscode-tools.svg)](https://marketplace.visualstudio.com/items?itemName=msazurermtools.azurerm-vscode-tools) [![Build Status](https://dev.azure.com/ms-azuretools/AzCode/_apis/build/status/vscode-azurearmtools)](https://dev.azure.com/ms-azuretools/AzCode/_build/latest?definitionId=10)

The Azure Resource Manager (ARM) Tools for Visual Studio Code provides language support, resource snippets, and resource auto-completion to help you create and validate Azure Resource Manager templates.

## Getting Started

The following guides detail how to use the ARM Tools extension features.

- [Snippets](#Snippets)
- [Azure schema integration](#Azure-schema-completion-and-validation)
- [Parameter file support](#parameter-files)
- [Completions](#completions)
- [Template navigation](#template-navigation)

> - [Other feature](#other-features)

Reviewer: ```suggestion
- [Other features](#other-features)
```

Author: Thanks, fixed.

- [Extension configuration](#extension-configuration)

## Snippets

Snippets are used to create basic templates and add elements such as parameters, resources, and outputs to existing templates.

### Create a new ARM template

When in an empty JSON file, typing `arm` produces a list of scaffolding snippets. These snippets can be used to create an empty template for any ARM deployment scope (Tenant, Subscription, Management Group, and Resource Group).

![Image showing the arm template scaffolding snippet](./images/json-snippets.png)

> Notice that adding scaffolding for an ARM template changes the Visual Studio Code language mode to *Azure Resource Manager Templates*.

Reviewer: ```suggestion
Notice that adding scaffolding for an ARM template changes the Visual Studio Code language mode to *Azure Resource Manager Template*.
```

Author: Fixed

![Image showing that the language type has changed to Azure Resource Manager Template](./images/arm-template-language.png)

### Add Azure resource to an ARM template

> Once you are working in an ARM template, placing your cursor in the resource and typing `arm` produces a list of 70+ snippets for Azure resources. Selecting a snippet adds an instance of the resource type to the ARM template.

Reviewer: ```suggestion
Once you are working in an ARM template, placing your cursor in the "resources" and typing `arm`, CTRL+SPACE or `{` produces a list of 70+ snippets for Azure resources. Selecting a snippet adds an instance of the resource type to the ARM template.
```

Similar changes needed in other places.  I'd like people to know about CTRL+SPACE (Mac or Windows) in general for completions since that seems to frequently be unknown to users.

Author: Fixed

> ![Image showing a storage account being added with a snippet](./images/arm-snippets.png)

Reviewer: Seems like an animation would be helpful here.  The term "tab stop" that vscode uses doesn't really explain anything by itself IMHO.

Author: fixed

Each snippet includes tab stops to help navigate through commonly modified configurations. Press the `tab` key to navigate to the next stop, and `shift` + `tab` to navigate to the previous stop. As seen in the following example, some stops are pre-populated with a list of common values for quick selection.

![Image showing snippet tab stops with pre-populated parameter values](./images/tab-stops.png)

### Snippet Reference

The following table lists the available ARM Tools snippets.

| Snippet | Description |

> |---|---|

Reviewer: Needs updating

Author: This is updated, please add any feedback / suggested changes.

### Parameter validation

Once a parameter file has been associated with a template, the extension provides parameter validation across both files. In the following example, a parameter is specified with a set of allowed values. In the parameter file, a value is specified that is not part of the allowed values list. Because of this, the extension produces an error.

![Image showing a parameter validation issue](./images/param-file-validation.png)

### Add missing parameters

To add missing parameters to an existing parameter file, ensure that an association has been created, click somewhere inside "_parameters_", and then click the light bulb icon.

![Image showing the light bulb helper for adding missing parameters](./images/missing-params-one.png)

Select the option to add only required or all missing parameters.

![Image showing missing parameter options (Add missing required parameters, Add all missing parameters)](./images/missing-params-two.png)

Enter a value for the added parameters and also remove the 'TODO' comment.

![Image showing parameter file after missing parameters have been added](./images/missing-params-three.png)

### Rename a parameter

> Right-click on the parameter name and select **Rename Symbol**. This action renames the parameter and all template references to the parameter.

Reviewer: ```suggestion
Right-click on the parameter name and select **Rename Symbol**. This action renames the parameter and all template and parameter file references to the parameter.
```

Author: Fixed

![Image showing the rename parameter dialoug](./images/rename-parameter.png)

### Remove or update parameter file association

Click on the code lens option for changing a parameter association.

![Image showing the code lens for creating a parameter file association](./images/undo-change-mapping.png)

Select a new parameter file or create a new parameter file to update the association. Select **None** to remove the parameter file association.

![Image showing the parameter file creation options (none, new, and browse)](./images/undo-change-mapping-two.png)

## Template navigation

The ARM tools extension for VS Code offers several features for navigating around an ARM template.

### ARM template outline

The ARM template outline is used to navigate to and select any element of an ARM template.

![Image showing the ARM template outline with a storage account resource selected](./images/arm-template-outline.png)

In addition to navigation, several features such as insert item and template sorting can be engaged using the buttons found on the ARM template outline.

### Go to Definition

> Go to definition can be used to navigate to both parameter and variable definitions. To use go to definition, place your cursor on the parameter or variable name and press `F12` or right-click and select **Go to Definition**.

Reviewer: ```suggestion
Go to definition can be used to navigate to parameter, variable and user-defined function definitions. To use go to definition, place your cursor on the parameter, variable or function name and press `F12` or right-click and select **Go to Definition**.
```

Author: Fixed

![Inserting an item into an Azure Resource Manager template](./images/insert-item.png)

You can also right-click on any element in the ARM Template Outline view to initiate the insert item process.

![Inserting an item into an Azure Resource Manager template](./images/insert-item-outline.png)

### Sort template

To sort elements of an ARM template alphabetically, right-click on the template and select **Sort Template...**.

![Image showing the sort template options (Resource, Function, Output, Parameter, Variable)](./images/sort-template.png)

Select the element that you want to sort.

![Image showing the insert resource menu](./images/sort-template-two.png)

Template sorting can also be engaged using the ARM template outline.

### Miscellaneous Features

  - [Signature help](https://code.visualstudio.com/docs/editor/editingevolved#_parameter-hints) for TLE function parameters
  - [Peek](https://code.visualstudio.com/docs/editor/editingevolved#_peek) for variable and parameter definitions
  - Find all references (Shift + F12) for variables and parameters

>   - Rename (F2) variables and parameters and their references

Reviewer: ```suggestion
  - Rename (F2) variables and parameters and their references, as well as user-defined functions or namespaces
```

Author: Fixed

  - [Hover](https://code.visualstudio.com/docs/editor/editingevolved#_hover) for parameter description
  - [TLE brace matching](https://code.visualstudio.com/docs/editor/editingevolved#_bracket-matching)
  - User-defined template functions, see Azure [documentation](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authoring-templates#functions)
  - Variable iteration ("copy blocks"), see Azure [documentation](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-multiple#variable-iteration)
  - Sort template and template sections alphabetically

> 

Reviewer: ```suggestion
  - Quick fix to add missing parameter values in the parameter file or in nested templates
```

Author: Fixed

| `armp!` | Adds the framework for a full deployment template parameters file. |
| `""` | While in the _parameters_ section adds a parameter to the template. |
| `""` | While in the _variables_ section adds a variable to the template. |

> | `{}` | While in the _functions_section add a new user-defined function. |

Reviewer: ```suggestion
| `{}` | While in the _functions_ section add a new user-defined function. |
```

| `arm-userfunc-namespace` | Adds a user function namespace to a template. |
| `arm-` | Displays 70+ snippets for creating Azure resources. For example, type `arm-ubuntu` to add all five resources necessary for a basic Ubuntu virtual machine. |

## Azure schema completion and validation

The ARM Tools extension uses Azure schemas to provide resource completion and validation. These schemas help you verify that Azure resources are configured correctly before deployment time.

### Schema Completions

> Entering double quotes `""` or selecting `ctrl` + `space` while the cursor is in a resource declaration provides a list of potential properties and property values for each Azure resource.

Reviewer: This applies to more than list resource declarations

Author: Fixed

In the following example, the potential values for an Azure Storage account tier are returned from the Azure Storage schema.

![Image showing a list of storage account tier values returned from the storage Azure schema](./images/completion.png)

### Schema Validation

> If an invalid property or property value is specified, a validation warning is thrown. In the following example, an invalid value is provided for the storage account tier (line 19). Notice that the warning message includes a list of valid values.

Reviewer: ```suggestion
If an invalid property or property value is specified, a validation warning is shown. In the following example, an invalid value has been provided for the storage account tier (line 19). Notice that the warning message includes a list of valid values.
```

![Image showing a schema validation error](./images/schema-validation.png)

## Parameter files

Associating a parameter file with an ARM template enables full validation across both the parameter file and template. A parameter file association can be made by either creating a new parameter file from an existing template or by associating an existing parameter file to the template.

### Create a new parameter file

> To create a new parameter file from an existing template, click on the code lense text `**Select or create a parameter file to enable full validation..**.

Reviewer: ```suggestion
To create a new parameter file from an existing template, click on the code lens text `**Select or create a parameter file to enable full validation..**.
```

Author: Fixed

![Image showing the code lense for creating a parameter file association](./images/create-parameter-file.png)

Select **New** from the context menu.

![Image showing the parameter file creation options (none, new, and browse)](./images/create-parameter-file-2.png)

Selecting **All parameters** adds every parameter from the template into the new parameter file. Selecting **Only required parameters** adds only parameters that do not have a default value specified in the template file.

![Image showing options for adding parameters for parameter file (All parameters, only required parameters)](./images/create-parameter-file-3.png)

Select a name and location for the parameter file.

![Image showing a file save as dialog box](./images/create-parameter-file-4.png)

> Once done, a parameter file has been created. Where needed, remove the comment indicating that a parameter value is needed and add a parameter value.

Reviewer: ```suggestion
Once done, a parameter file is been created. Where needed, remove the comment indicating that a parameter value is needed and add a parameter value.
```

Reviewer: Too many "needed"s in this sentence... :-)

Author: Fixed

![Image showing new parameter file](./images/parameter-file.png)

> Also, notice that on the template file that the code lense text has been updated to reflect the current parameter file association.

Reviewer: ```suggestion
Also, notice that in the template file that the code lens text has been updated to reflect the current parameter file association.
```

Reviewer: Probably also point out the status bar.

Author: Fixed

> ![Image showing that the parameter code lens has been updated to reflect associated parameter file](./images/code-lense.png)

Reviewer: nit: lense => lens in file name.

Please correct "lense" elsewhere in the readme, thanks.

Author: Fixed

### Associate an existing parameter file

> To create an association with an existing parameter file, click on the code lense text `**Select or create a parameter file to enable full validation..**.

Reviewer: extra period ".." and missing the closing backtick

Author: Fixed

![Image showing the code lense for creating a parameter file association](./images/create-parameter-file.png)

> Select **Browse** from the context menu and then select the parameter file.

Reviewer: If it's in the same folder, browse isn't necessary.

Author: Fixed

![Image showing the parameter file creation options (none, new, and browse)](./images/create-parameter-file-2.png)

### Parameter validation

Once a parameter file has been associated with a template, the extension provides parameter validation across both files. In the following example, a parameter is specified with a set of allowed values. In the parameter file, a value is specified that is not part of the allowed values list. Because of this, the extension produces an error.

> ![Image showing a parameter validation issue](./images/param-file-validation.png)

Reviewer: Very ugly error from validation, do we have a simpler example?

Author: This error is demonstrating that a value has been specified in the parameters file that is not part of the allowed values list. I've looked at a validation error for a value that exceeds max value and the error is equally as ugly. I'm not sure what parameter validation would have a more simple error, if you can provide one, I will update.

### Add missing parameters

> To add missing parameters to an existing parameter file, ensure that an association has been created, select `parameters`, and then click the light bulb icon.

Reviewer: ```suggestion
To add missing parameters to an existing parameter file, ensure that an association has been created, click somewhere inside `"parameters"`, and then click the light bulb icon.
```

Author: Fixed

![Image showing the light bulb helper for adding missing parameters](./images/missing-params-one.png)

Select the option to add only required or all missing parameters.

![Image showing missing parameter options (Add missing required parameters, Add all missing parameters)](./images/missing-params-two.png)

Enter a value for the added parameters and also remove the 'TODO' comment.

![Image showing parameter file after missing parameters have been added](./images/missing-params-three.png)

> 

Reviewer: Since most of these work for parameters, variables, user namespaces and user functions, I'd suggest having a separate section for those, and then just point to those sections from here, rather than only mention them here in relation to parameters.

Author: Fixed

![Image showing the rename parameter dialoug](./images/rename-parameter.png)

### Remove or update parameter file association

Click on the code lense option for changing a parameter association.

![Image showing the code lense for creating a parameter file association](./images/undo-change-mapping.png)

Select a new parameter file or create a new parameter file to update the association. Select **None** to remove the parameter file association.

![Image showing the parameter file creation options (none, new, and browse)](./images/undo-change-mapping-two.png)

## Template navigation

The ARM tools extension for VS Code offers several features for navigating around an ARM template.

### ARM template outline

The ARM template outline is used to navigate to and select any element of an ARM template.

![Image showing the ARM template outline with a storage account resource selected](./images/arm-template-outline.png)

In addition to navigation, several features such as insert item and template sorting can be engaged using the buttons found on the ARM template outline.

> ### Go to Definition

Reviewer: I'd reference Find Definitions here as well.

Author: I've added a find all references just below. Is this what you mean by find definition?

![Image showing the Go to Definition dialog on a parameter reference](./images/go-to-definition.png)

### Find all references

Right-click on a parameter, variable, or user-defined function name and select **Go to References**. Double-clicking on any reference moves your cursor to the reference location.

![Image showing the go to reference user interface](./images/parameter-references.png)

## Completions

The ARM Tools extension provides code completions for the following:

>   - [Built-in functions such as resourceId and variable](https://go.microsoft.com/fwlink/?LinkID=733958)

Reviewer: variable?  Vars/params are mentioned specially in the next couple of lines.

Author: Fixed, thanks.

>   - [resourceId() function](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource#resourceid)

Reviewer: I'd say built-in functions and give resourceId/etc as examples.

Maybe also say resourceId arguments separately?

Author: Fixed

  - Properties of references to variables that are objects

> ## Other feature

Reviewer: ```suggestion
## Other features
```

Author: Fixed

### Insert Item

> In addition to adding snippets from the code editor, the Insert Item feature can be used to insert new parameters, user-defined functions, variables, resources, and outputs. To do so right-click on the template in the code editor, select **Insert Item** and follow the on-screen prompts.

Reviewer: ```suggestion
In addition to adding snippets from the code editor, the Insert Item feature can be used to insert new parameters, user-defined functions, variables, resources, and outputs. To do so right-click the code editor, select **Insert Item** and follow the on-screen prompts.
```

Author: Fixed

You may be interested in adjusting the following extension configurations. These can be configured in [VS Code User Settings](https://code.visualstudio.com/docs/getstarted/settings).

| Configuration | Description |
|---|---|

> | azureResourceManagerTools.autoDetectJsonTemplates | Auto-detects ARM template and sets editor language type to Azure Resource Manager. |

Reviewer: ```suggestion
| azureResourceManagerTools.autoDetectJsonTemplates | Auto-detects ARM template and sets editor language type to Azure Resource Manager Template. |
```

| azureResourceManagerTools.checkForLatestSchema | Check if the root schema for deployment templates is using an out-of-date version and suggest updating it to the latest version. |

> | azureResourceManagerTools.checkForMatchingParameterFiles | Check if an opened template file has a matching params file and prompt to create an association. |

Reviewer: ```suggestion
| azureResourceManagerTools.checkForMatchingParameterFiles | Check if an opened template file has a matching parameter file and prompt to create an association. |
```

> | azureResourceManagerTools.codelens.enable | Disable all code lens functionality. |

Reviewer: ```suggestion
| azureResourceManagerTools.codelens.enable | Set to false to disable all code lens functionality. |
```

> | azureResourceManagerTools.codelens.parameters | Disable code lens for parameters. |

Reviewer: ```suggestion
| azureResourceManagerTools.codelens.parameters | Set to false to disable code lens for parameters. |
```

  - Variable iteration ("copy blocks"), see Azure [documentation](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-multiple#variable-iteration)
  - Sort template and template sections alphabetically

> ## Extension configuration

Reviewer: I think these would be much better in a short table format - these take up too much space for relatively minor features and make the readme look more complex and confusing IMHO.

Author: Fixed.

```

Parameter file assoications are stored in the `azureResourceManagerTools.parameterFiles` user setting.

```
"azureResourceManagerTools.parameterFiles": {}
```

### Code lense

Enable or disable all code lens features.

```
"azureResourceManagerTools.codelens.enable": true,
```

Enable code lens for parameter definitions and values.

```
 "azureResourceManagerTools.codelens.parameters": true,
```

## Automatic Detection of deployment template files

By default, the extension recognizes a .json or .jsonc file as a deployment template file based on the \$schema specified in the file (for instance, `https://schema.management.azure.com/schemas/2018-05-01/deploymentTemplate.json#`) and will switch the editor language to "Azure Resource Manager Template" automatically. If you do not want that behavior, you can set the `azureResourceManagerTools.autoDetectJsonTemplates` setting to false and use the below methods to determine which files to treat as deployment templates.

Besides automatic detection, you can also use the `files.associations` setting to set up your own specific associations based on specific files paths or patterns to mark them as deployment templates, e.g.

```json

> "files.associations": {

Reviewer: Need to update the troubleshooting section unrecognized api version so that it contains the actual error message (so people can search for it).

Author: Fixed

}
```

> Auto-detects ARM template and sets editor language type to Azure Resource Manager Template

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

## Troubleshooting

Use the following wiki article to help troubleshoot these known issues.

> - [Could not find/download: ".NET Core Runtime" with version = 2.2.5](https://github.com/microsoft/vscode-azurearmtools/wiki/Troubleshooting-DotNet-Acquisition)

Reviewer: ```suggestion
- [This extension requires .NET Core for full functionality, but we were unable to download and install a local copy for the extension.](https://github.com/microsoft/vscode-azurearmtools/wiki/Troubleshooting-DotNet-Acquisition)
```

Author: Fixed

> - [Resource schema warning: "For full schema validation, consider updating the value to one of the following.."](https://github.com/microsoft/vscode-azurearmtools/wiki/Unrecognized-Resource-API-Version)

Reviewer: ```suggestion
- [Resource schema warning: "For full schema validation, consider updating the value to one of the following..."](https://github.com/microsoft/vscode-azurearmtools/wiki/Unrecognized-Resource-API-Version)
```

Author: Fixed
